docs/blueprints/EP031.rst

```diff
@@ -0,0 +1,273 @@
+    - Italo Valcy <idasilva AT fiu DOT edu>
+    - Vinicius Arcanjo <vindasil AT fiu DOT edu>
+:Created: 2022-08-24
+:Kytos-Version: 2022.3
+:Status: Draft
+
+****************************************
+EP031 - Telemetry-enabled EVCs
+****************************************
+
+
+Abstract
+========
+
+Blueprint 31 presents requirements and features to support In-band Network Telemetry (a.k.a. INT or telemetry) for EVCs created by the Kytos **mef-eline** napp.
+
+
+Motivation
+==========
+
+INT enables per-packet monitoring and it is specified by the P4.org consortium as one of P4 use cases. To enable full network visibility, all EVCs must support INT. For the network devices available at AmLight, INT will be performed by the network data plane when instructed to do so via OpenFlow NoviFlow Experimenter's actions.
+
+Supporting INT with the current implementation of the NoviFlow's NoviWare OS won't be trivial due to many conditions imposed by NoviWare. This blueprint will focus on enabling a MINIMALLY functional support for INT over Kytos. Once this blueprint is implemented, new blueprints will be created addressing more complex requirements.
+
+
+I. Requirements
+===============
+
+This blueprint has the following characteristics:
+
+  1. There will be no concerns about number or location of the INT Hops switches. Currently, an INT Sink switch can only export up to 10 metadata stacked by the network.
+  2. There will be no concerns about MTU restrictions. NoviWare and Tofino support frames as large as 10KB which is more than enough. However, any legacy device in the middle must be previously identified since legacy devices usually have a MTU of up to 9216 Bytes.
+  3. There will be no concerns if INT Sink switches have loops for INT reports (Section III). This blueprint assumes physical loops are already deployed by the operators during the commissioning phase.
+  4. There will be no concerns about proxy ports and their mappings.
+  5. There is no need for persistent data. The **mef_eline** and **flow_manager** napps will persists their entries accordingly since **telemetry** will leverage **flow_manager**.
+  6. This version won't require changes to the way the **mef_eline** napp works.
+
+
+II. How INT works with NoviWare
+===============================
+
+Currently, the **mef_eline** napp creates Ethernet Virtual Circuits (EVCs) using *matches* `IN_PORT` and `VLAN_VID` and actions `PUSH/POP/SWAP VLAN`, `SET FIELD`, and `OUTPUT`. All EVCs' flow entries have the same priority. Currently, INT is only supported for IPv4 traffic with TCP and UDP protocols, which means that more specific matches will be needed to match IP and TCP/UDP. Non-IPV4 or non-TCP/UDP traffic will continue using the existing flow entries created by **mef_eline**. The **telemetry** napp only adds and removes flows for INT. Any other flows created by the **mef_eline** won't be affected. Example:
+
+- Current by **mef_eline**:
+
+  - 1 flow per direction
```

Comment:
Good catch! Can you provide an example? I am not familiar with the failover path.

Reply:
yes, sure! From [EP029](https://github.com/kytos-ng/kytos/blob/master/docs/blueprints/EP029.rst):

Suppose that the primary path will be set up to S - X - W - D. The maximum disjoint path will be set up to S - Z - W - I - D. The following rules will be created before a link failure:

Rules for the primary path:
```
h1 -> S -> X -> W -> D -> h2
h2 -> D -> W -> X -> S -> h1
```

Rules for the failover path:
```
S -> Z -> W -> I -> D -> h2
D -> I -> W -> Z -> S -> h1

```

Reply:
I created section IX to address this failover integration. In the end, since the telemetry napp will modify all flows based on cookie ID, INT will be enabled in the failover path. Thanks.